File: easy/additive_persistence.py

def nonbonus_additive_persistence(number):
    number = [int(i) for i in str(number)]
    count = 0
    while len(number) > 1:
        number = [int(i) for i in str(sum(number))]
        count +=1
    return count


def decompose(number):
    number_repr = []
    digits = len(str(number))
    for i in reversed(range(digits)):
        number_repr.append(number // 10**i)
        number = number -  (10**i)* number_repr[-1]
    return number_repr


def bonus_additive_persistence(number):
    count  = 0
    while number >= 10:
        number = sum(decompose(number))
        count += 1
    return count





print(nonbonus_additive_persistence(13))
print(nonbonus_additive_persistence(1234))
print(nonbonus_additive_persistence(9876))
print(nonbonus_additive_persistence(199))

print(bonus_additive_persistence(13))
print(bonus_additive_persistence(1234))
print(bonus_additive_persistence(9876))
print(bonus_additive_persistence(199))

File: easy/additive_persistence.cpp
#include <iostream>
#include <vector>
#include <cmath>

std::string print_vector(std::vector<int> myvector)
{
    std::string mystring;
    for (int i : myvector)
        mystring += std::to_string(i);
    mystring += "\n";
    return mystring;
}

int decompose_and_sum(int number)
{
    std::vector<int> number_repr;
    int digits = std::to_string(number).size() - 1;
    for (int i = digits; i >= 0; i--)
    {
        number_repr.push_back(number / std::pow(10, i));
        number -= (std::pow(10, i) * number_repr.back());
    }
    int sum = 0;
    for (int i : number_repr)
        sum += i;
    return sum;
}

int bonus_additive_persistence(int number)
{
    int count = 0;
    while (number >= 10)
    {
        int newnumber = decompose_and_sum(number);
        number = newnumber;
        count += 1;
    }
    return count;
}

int main()
{
    std::cout << bonus_additive_persistence(13);
    std::cout << bonus_additive_persistence(1234);
    std::cout << bonus_additive_persistence(9876);
    std::cout << bonus_additive_persistence(199);
}
